### ISSUE TITLE
Systems which aren't part of a SystemSet should be configurable after having been added to ScheduleGraph

### ISSUE THREAD

**Issue by MeoMix** [bevyengine/bevy#11970]
## What problem does this solve or what need does it fill?

I have several systems which I run within the `FixedUpdate` schedule. These systems depend on resources which don't exist during `Startup`, but do exist shortly thereafter. I use `State` to indicate when all of my resources have been inserted so that systems do not need to expect optional arguments.

I am able to achieve this by writing:

```rust
app.add_systems(FixedUpdate, system_a.run_if(AppState::Running));
...
app.add_systems(FixedUpdate, system_b.run_if(AppState::Running));
...
app.add_systems(FixedUpdate, system_c.run_if(AppState::Running));
```

However, I would prefer to not have to repeat the same run condition in each location.

One way to achieve this would be to apply a run condition to `run_fixed_update_schedule` which is added as part of `TimePlugin`:

```rust
impl Plugin for TimePlugin {
    fn build(&self, app: &mut App) {
        app.init_resource::<Time>()
            .init_resource::<Time<Real>>()
            .init_resource::<Time<Virtual>>()
            .init_resource::<Time<Fixed>>()
            .init_resource::<TimeUpdateStrategy>()
            .register_type::<Time>()
            .register_type::<Time<Real>>()
            .register_type::<Time<Virtual>>()
            .register_type::<Time<Fixed>>()
            .register_type::<Timer>()
            .register_type::<Stopwatch>()
            .add_systems(
                First,
                (time_system, virtual_time_system.after(time_system)).in_set(TimeSystem),
            )
            .add_systems(RunFixedUpdateLoop, run_fixed_update_schedule);
```

Unfortunately, this is either difficult or impossible with the current codebase. If I write:

```rust
app.add_systems(RunFixedUpdateLoop, run_fixed_update_schedule.run_if(in_state(AppState::Running));
```

the run condition is not respected.

In contrast, if a SystemSet were to be used then this limitation does not exist. As an example:

```rust
#[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
pub struct TestSystemSet;

fn test_system() {
    info!("running");
}

app.add_systems(
    Update,
    test_system.in_set(TestSystemSet),
);

app.configure_sets(Update, TestSystemSet.run_if(in_state(AppState::Running)));
```

Here, `test_system` does not log `running` even though the system was initially added without a run condition.

## What solution would you like?

I would like individual systems to have the same configurability as system sets. My uninformed perspective is that these two approaches should functionally identically except that SystemSet allows me to refer to multiple systems as a group. 

## What alternative(s) have you considered?

For this issue, I think that I can load TimePlugin after my own code and ensure that my code, including run condition, takes precedent.

**Comment by alice-i-cecile** [comment]
@maniwani @hymm @JoJoJet do any of you know why this isn't currently possible? Was this a deliberate choice? Do you feel strongly about whether this should be allowed or not?

**Comment by hymm** [comment]
It's potentially a footgun if you have multiple of the same system in the same schedule. i.e. add a run criteria to apply_deferred. If we can relax it in a reasonable way. i.e. panic if you try to configure multiple off the same system than I think it would be ok. I think it's might not be possible to detect that at configuration time though. We could maybe do during schedule construction, but would be hard to point to where the failure is occuring.

**Comment by alice-i-cecile** [comment]
Yeah, I'd be happy with a panic here.

**Comment by maniwani** [comment]
If the problem is "I'd like a less boilerplate way to condition a group of systems" I don't understand how that takes us to "I need to condition the schedule itself".

IMO you identified the appropriate solution: Define your own system set and use it.

> ```rust
> #[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
> pub struct TestSystemSet;
> 
> app.add_systems(
>     FixedUpdate,
>     test_system.in_set(TestSystemSet),
> );
> 
> app.configure_sets(Update, TestSystemSet.run_if(in_state(AppState::Running)));
> ```


> I would like individual systems to have the same configurability as system sets.

> Do any of you know why this isn't currently possible?

Unfortunately, it's prone to error. Please see [my comment][1] in #11971. You can see how adding a condition to the `TimePlugin`'s logic easily breaks assumptions it was making.

[1]: https://github.com/bevyengine/bevy/pull/11971#issuecomment-1951618032

**Comment by MeoMix** [comment]
> If the problem is "I'd like a less boilerplate way to condition a group of systems" I don't understand how that takes us to "I need to condition the schedule itself".
> 
> IMO you identified the appropriate solution: Define your own system set and use it.
> 
> > ```rust
> > #[derive(Debug, Hash, PartialEq, Eq, Clone, SystemSet)]
> > pub struct TestSystemSet;
> > 
> > app.add_systems(
> >     FixedUpdate,
> >     test_system.in_set(TestSystemSet),
> > );
> > 
> > app.configure_sets(FixedUpdate, TestSystemSet.run_if(in_state(AppState::Running)));
> > ```
> 
> > I would like individual systems to have the same configurability as system sets.
> 
> > Do any of you know why this isn't currently possible?
> 
> Unfortunately, it's prone to error. Please see [my comment](https://github.com/bevyengine/bevy/pull/11971#issuecomment-1951618032) in #11971. You can see how adding a condition to the `TimePlugin`'s logic easily breaks assumptions it was making.

I am not able to define a SystemSet for third party code, though, correct?

I feel the error prone concern should be addressed with proper visibility modifiers. It seems like an equally concerning footgun to say that a public system, which happens to be part of a set, is now configurable. If the system isn't configurable then it shouldn't be exposed irrespective of whether it's part of a set.

**Comment by MeoMix** [comment]
> It's potentially a footgun if you have multiple of the same system in the same schedule. i.e. add a run criteria to apply_deferred. If we can relax it in a reasonable way. i.e. panic if you try to configure multiple off the same system than I think it would be ok. I think it's might not be possible to detect that at configuration time though. We could maybe do during schedule construction, but would be hard to point to where the failure is occuring.

Yeah, I can see how this would be a concern. As I think about `apply_deferred`, perhaps I'm also misunderstanding how things are supposed to work.

My intent was to say "I know this system has already been added to a schedule. I want to modify that system." but clearly it's possible to add a system with the same name twice to a schedule - that's how `apply_deferred` works.

It sounds like what I was actually doing was adding a second instance of the system to the schedule and potentially decorating both instances with a `run_if` condition. I think that's the only way this could be implemented while acknowledging how `apply_deferred` works.

So, I suppose it is a footgun as I was bumping into it myself without realizing it? :)

Anyway, I think we can shelve this idea for now.

**Comment by maniwani** [comment]
> I am not able to define a SystemSet for third party code, though, correct?

> What I saw in code was that add_systems takes an `impl IntoSystemConfigs` and `configure_sets` takes an `impl IntoSystemSetConfigs`. This led me to strongly believe that I should be able to configure systems being passed into `add_systems` as the argument interface appeared nearly identical aside from the implication of a group vs individual.

Hmm, it's complicated and not all that intentional.

The reason for the difference is simply because, inside a schedule, system sets are unique and systems aren't. A system set can remain configurable at all times simply because we can always look it up. On the other hand, many instances of system `F` may have been added to a schedule, so there's no guarantee that `F` uniquely identifies a specific instance.

I think this is worth reconsidering now that the insertion of `apply_deferred` (the main system with duplicates) is automated.

> I feel the error prone concern should be addressed with proper visibility modifiers. It seems like an equally concerning footgun to say that a public system, which happens to be part of a set, is now configurable. If the system isn't configurable then it shouldn't be exposed irrespective of whether it's part of a set.

Easier said than done. To give some context, IIRC @alice-i-cecile, @cart, and I did try to brainstorm ways to enforce visibility/permissions for system sets when we were finalizing the "stageless" RFC, but I don't remember us landing on one. Like how would a plugin know which consumers have the necessary permissions to mess with its system sets?

These days, I'm more convinced that visibility/permissions just aren't something the engine itself should (or even _can_) do.

But yeah, just because the API can't stop the user from breaking things if they try doesn't mean that it's encouraged. IMO users should be extra careful about configuring systems and sets that aren't theirs. I think the main reason plugins make their system sets publicly visible is so that downstream users can order *around* those sets, not so much modify them.
